### PR TITLE
PiPPy + TP

### PR DESCRIPTION
**Pull request by kwen2501**
### Test command:
```
torchrun --nproc_per_node=8 pippy_tp.py
```
- Requires PyTorch 2.0 for DistributedTensor and Tensor Parallel support.
- World size = 8
- Number of pipeline stages = 4
- Tensor parallel group size = 2

### Output
```
GSPMD rank 5, PP ranks: [1, 3, 5, 7], my rank in pipe: 2
GSPMD rank 2, PP ranks: [0, 2, 4, 6], my rank in pipe: 1
GSPMD rank 4, PP ranks: [0, 2, 4, 6], my rank in pipe: 2
GSPMD rank 3, PP ranks: [1, 3, 5, 7], my rank in pipe: 1
GSPMD rank 7, PP ranks: [1, 3, 5, 7], my rank in pipe: 3
GSPMD rank 1, PP ranks: [1, 3, 5, 7], my rank in pipe: 0
GSPMD rank 0, PP ranks: [0, 2, 4, 6], my rank in pipe: 0
GSPMD rank 6, PP ranks: [0, 2, 4, 6], my rank in pipe: 3

GraphModule(
  (submod_0): GraphModule(
    (mlp0_net1): Linear(in_features=256, out_features=256, bias=True)
    (mlp0_relu): ReLU()
    (mlp0_net2): Linear(in_features=256, out_features=256, bias=True)
  )
  (submod_1): GraphModule(
    (mlp1_net1): Linear(in_features=256, out_features=256, bias=True)
    (mlp1_relu): ReLU()
    (mlp1_net2): Linear(in_features=256, out_features=256, bias=True)
  )
  (submod_2): GraphModule(
    (mlp2_net1): Linear(in_features=256, out_features=256, bias=True)
    (mlp2_relu): ReLU()
    (mlp2_net2): Linear(in_features=256, out_features=256, bias=True)
  )
  (submod_3): GraphModule(
    (mlp3_net1): Linear(in_features=256, out_features=256, bias=True)
    (mlp3_relu): ReLU()
    (mlp3_net2): Linear(in_features=256, out_features=256, bias=True)
  )
)



def forward(self, x):
    submod_0 = self.submod_0(x);  x = None
    submod_1 = self.submod_1(submod_0);  submod_0 = None
    submod_2 = self.submod_2(submod_1);  submod_1 = None
    submod_3 = self.submod_3(submod_2);  submod_2 = None
    return submod_3
    

Rank 1 calling parallelize_module with DeviceMesh:([0, 1])
Rank 2 calling parallelize_module with DeviceMesh:([2, 3])
Rank 6 calling parallelize_module with DeviceMesh:([6, 7])
Rank 0 calling parallelize_module with DeviceMesh:([0, 1])
Rank 4 calling parallelize_module with DeviceMesh:([4, 5])
Rank 3 calling parallelize_module with DeviceMesh:([2, 3])
Rank 7 calling parallelize_module with DeviceMesh:([6, 7])
Rank 5 calling parallelize_module with DeviceMesh:([4, 5])

Rank 0 Instantiated pipeline with ranks [0, 2, 4, 6]
Rank 1 Instantiated pipeline with ranks [1, 3, 5, 7]

Pipeline 1 output: torch.Size([32, 256])
Pipeline 0 output: torch.Size([32, 256])
```

### CODE REVIEW

File: examples/tp+pp/pippy_tp.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        pipe_split()
+        x = self.mlp1(x)
+        pipe_split()
+        x = self.mlp2(x)
+        pipe_split()
+        x = self.mlp3(x)
+        return x
+
+
+def run_gspmd(pp_ranks, args):
```
Comment (fduwjj) [inline examples/tp+pp/pippy_tp.py:72]: shall we use other name rather than "gspmd"?

**Review by fduwjj — APPROVED** [review]
Amazing work! Thanks!!

File: examples/tp+pp/pippy_tp.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    # PiPPy tracing
+    ec_pipe = Pipe.from_tracing(ec)
+    if args.rank == 0:
+        print(ec_pipe.split_gm)
+
+    # Ask PiPPy to export stage submodule
+    submod = ec_pipe.export(stage_id=pp_rank)
+
+    # Create TP device mesh
+    my_device_mesh = None
```
Comment (kumpera) [inline examples/tp+pp/pippy_tp.py:120]: This should be using a single global 2D DeviceMesh with one dimension assigned for PP and the other for TP.

Comment (kwen2501) [reply examples/tp+pp/pippy_tp.py:120]: Great callout @kumpera.
It requires PiPPy to support `DeviceMesh` first. I will look into that soon.

**Pull request merged**